File: mqtt/solar/anomaly.py
"""異常偵測（Feature #5）。

規則：
- 白天時段（由 schedule.is_daytime 判斷），若總功率為 0 或 None 持續超過
  `daytime_zero_minutes` 分鐘，視為異常。
- 單區功率為 0 持續超過門檻，逐區觸發。
- 觸發後每次持續仍為 0 不重複 alert（僅首次），恢復後才會 reset。
"""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timedelta
from typing import Callable


@dataclass
class _ZeroWindow:
    start: datetime | None = None
    fired: bool = False

    def update(self, is_zero: bool, now: datetime, threshold: timedelta) -> str:
        """回傳事件：'start', 'fire', 'recover', 'none'。"""
        if is_zero:
            if self.start is None:
                self.start = now
                self.fired = False
                return "start"
            if not self.fired and (now - self.start) >= threshold:
                self.fired = True
                return "fire"
            return "none"
        # 非零 → 恢復
        if self.start is not None:
            event = "recover" if self.fired else "none"
            self.start = None
            self.fired = False
            return event
        return "none"


@dataclass
class AnomalyDetector:
    daytime_zero_minutes: int = 5
    on_alert: Callable[[str, str, str], None] | None = None  # (factory, level, msg)

    _factory_windows: dict[str, _ZeroWindow] = field(default_factory=dict)
    _zone_windows: dict[tuple[str, int], _ZeroWindow] = field(default_factory=dict)

    def check(
        self,
        factory_id: str,
        summary: dict,
        zones: list[dict],
        is_daytime: bool,
        now: datetime | None = None,
    ) -> None:
        now = now or datetime.now()
        threshold = timedelta(minutes=self.daytime_zero_minutes)

        if not is_daytime:
            # 夜間 reset
            self._factory_windows.pop(factory_id, None)
            for k in [k for k in self._zone_windows if k[0] == factory_id]:
                self._zone_windows.pop(k, None)
            return

        # 全廠
        total = summary.get("total_power_kw")
        fw = self._factory_windows.setdefault(factory_id, _ZeroWindow())
        ev = fw.update(_is_zero(total), now, threshold)
        self._emit(factory_id, "factory", None, ev, total, fw, threshold)

        # 各區
        for z in zones:
            zid = z["zone_id"]
            key = (factory_id, zid)
            zw = self._zone_windows.setdefault(key, _ZeroWindow())
            power = z.get("power_kw")
            ev = zw.update(_is_zero(power), now, threshold)
            self._emit(factory_id, "zone", z, ev, power, zw, threshold)

    def _emit(
        self,
        factory_id: str,
        scope: str,
        zone: dict | None,
        event: str,
        value,
        window: _ZeroWindow,
        threshold: timedelta,
    ) -> None:
        if event == "none" or self.on_alert is None:
            return
        if scope == "factory":
            target = "全廠"
        else:
            zid = zone["zone_id"] if zone else "?"
            name = (zone or {}).get("name") or ""
            target = f"Zone{zid} {name}".strip()

        if event == "fire":
            msg = f"{target} 白天連續零功率超過 {threshold.total_seconds()/60:.0f} 分鐘"
            self.on_alert(factory_id, "WARN", msg)
        elif event == "recover":
            v = f"{value:.1f} kW" if isinstance(value, (int, float)) else "有值"
            msg = f"{target} 已恢復發電（{v}）"
            self.on_alert(factory_id, "INFO", msg)


def _is_zero(v) -> bool:
    if v is None:
        return True
    try:
        return float(v) <= 0.0
    except (ValueError, TypeError):
        return True

File: solar_mqtt_go/internal/anomaly/anomaly.go
// Package anomaly 實作白天零功率異常偵測（對應 Python solar/anomaly.py）。
//
// 規則：
//   - 白天時段（由呼叫端依 night_pause 設定傳入 is_daytime），若總功率為 0 或
//     缺值持續超過 daytime_zero_minutes，視為異常（WARN，僅首次）。
//   - 單區功率為 0 同步觸發；恢復（非零）時發 INFO，帶當前值。
//   - 夜間（is_daytime=false）重置該廠所有 window。
package anomaly

import (
	"fmt"
	"strings"
	"sync"
	"time"

	"solar_mqtt_go/internal/scraper"
)

// IsZero 對應 Python _is_zero：nil、<=0 均視為 zero。
func IsZero(v *float64) bool {
	if v == nil {
		return true
	}
	return *v <= 0.0
}

// Window 零功率累積視窗（對應 Python _ZeroWindow）。
type Window struct {
	start *time.Time
	fired bool
}

// Update 更新視窗狀態，回傳事件："start"、"fire"、"recover"、"none"。
func (w *Window) Update(isZero bool, now time.Time, threshold time.Duration) string {
	if isZero {
		if w.start == nil {
			t := now
			w.start = &t
			w.fired = false
			return "start"
		}
		if !w.fired && now.Sub(*w.start) >= threshold {
			w.fired = true
			return "fire"
		}
		return "none"
	}
	// 非零 → 恢復
	if w.start != nil {
		event := "none"
		if w.fired {
			event = "recover"
		}
		w.start = nil
		w.fired = false
		return event
	}
	return "none"
}

// AlertFn 告警回呼（factory_id, level, message）。
type AlertFn func(factoryID, level, message string)

type zoneKey struct {
	factory string
	zoneID  int
}

// Detector 異常偵測器（對應 Python AnomalyDetector）。
type Detector struct {
	DaytimeZeroMinutes int
	onAlert            AlertFn

	mu             sync.Mutex
	factoryWindows map[string]*Window
	zoneWindows    map[zoneKey]*Window
}

// New 建立 detector。
func New(daytimeZeroMinutes int, onAlert AlertFn) *Detector {
	return &Detector{
		DaytimeZeroMinutes: daytimeZeroMinutes,
		onAlert:            onAlert,
		factoryWindows:     map[string]*Window{},
		zoneWindows:        map[zoneKey]*Window{},
	}
}

// SetDaytimeZeroMinutes 熱更新門檻（/set 變更）。
func (d *Detector) SetDaytimeZeroMinutes(minutes int) {
	d.mu.Lock()
	defer d.mu.Unlock()
	d.DaytimeZeroMinutes = minutes
}

// Check 檢查一輪讀值。isDaytime=false 時重置該廠所有 window。
func (d *Detector) Check(factoryID string, summary *scraper.Summary, zones []scraper.Zone, isDaytime bool, now time.Time) {
	d.mu.Lock()
	threshold := time.Duration(d.DaytimeZeroMinutes) * time.Minute
	d.mu.Unlock()

	if !isDaytime {
		// 夜間 reset（對應 Python pop）
		d.mu.Lock()
		delete(d.factoryWindows, factoryID)
		for k := range d.zoneWindows {
			if k.factory == factoryID {
				delete(d.zoneWindows, k)
			}
		}
		d.mu.Unlock()
		return
	}

	var total *float64
	if summary != nil {
		total = summary.TotalPowerKw
	}

	d.mu.Lock()
	fw := d.factoryWindows[factoryID]
	if fw == nil {
		fw = &Window{}
		d.factoryWindows[factoryID] = fw
	}
	d.mu.Unlock()
	ev := fw.Update(IsZero(total), now, threshold)
	d.emit(factoryID, "全廠", ev, total, threshold)

	for _, z := range zones {
		key := zoneKey{factoryID, z.ZoneID}
		d.mu.Lock()
		zw := d.zoneWindows[key]
		if zw == nil {
			zw = &Window{}
			d.zoneWindows[key] = zw
		}
		d.mu.Unlock()
		ev := zw.Update(IsZero(z.PowerKw), now, threshold)
		target := strings.TrimSpace(fmt.Sprintf("Zone%d %s", z.ZoneID, z.Name))
		d.emit(factoryID, target, ev, z.PowerKw, threshold)
	}
}

// emit 對應 Python _emit：fire → WARN、recover → INFO（帶當前值）。
func (d *Detector) emit(factoryID, target string, event string, value *float64, threshold time.Duration) {
	if event == "none" || event == "start" || d.onAlert == nil {
		return
	}
	switch event {
	case "fire":
		msg := fmt.Sprintf("%s 白天連續零功率超過 %.0f 分鐘", target, threshold.Minutes())
		d.onAlert(factoryID, "WARN", msg)
	case "recover":
		v := "有值"
		if value != nil {
			v = fmt.Sprintf("%.1f kW", *value)
		}
		msg := fmt.Sprintf("%s 已恢復發電（%s）", target, v)
		d.onAlert(factoryID, "INFO", msg)
	}
}
